### PR TITLE
Admin can update Membership Application status

### PR DESCRIPTION
**Pull request by thesuss**
PT Story: https://www.pivotaltracker.com/story/show/133950603

_As an Admin
In order to get members into SHF and get their money
I need to be able to accept/reject their application_

Changes proposed in this pull request:
1. Add `status` to Membership Application (default 'pending')
2. Adds select box to Membership Application show page
3. Sets select box at current status of Membership Application (to be refactored)
4. Show timestamp and note of when it was accepted

### CODE REVIEW

File: db/migrate/20161112171216_add_status_to_membership_application.rb
Comment (thesuss) [inline db/migrate/20161112171216_add_status_to_membership_application.rb]: So, Single quotes except Cucumber, string interpolation and Migrations ?

File: features/step_definitions/application_steps.rb
Comment (thesuss) [inline features/step_definitions/application_steps.rb]: How do I do the same in a different way? @AgileVentures/shf-project-team

File: db/migrate/20161112171216_add_status_to_membership_application.rb
Comment (thesuss) [reply db/migrate/20161112171216_add_status_to_membership_application.rb:1]: Don't understand this one, even after googling...

File: app/controllers/memberships_controller.rb
Comment (tochman) [reply app/controllers/memberships_controller.rb]: Why is the update method here. It is part of another PR. You need to keep track of what branch you branch off from. This may lead to future merge conflicts.

File: app/controllers/memberships_controller.rb
Comment (tochman) [reply app/controllers/memberships_controller.rb]: Not introduced by me. but still part of another PR

File: config/routes.rb
Comment (tochman) [inline config/routes.rb]: Also part of another PR

File: db/schema.rb
Comment (tochman) [inline db/schema.rb:46]: ??

File: features/step_definitions/application_steps.rb
Comment (tochman) [reply features/step_definitions/application_steps.rb]: This step def is generic. I would place it in `basic_steps`

File: features/step_definitions/application_steps.rb
Comment (tochman) [inline features/step_definitions/application_steps.rb]: This step def is generic. I would place it in basic_steps

File: features/step_definitions/application_steps.rb
Comment (tochman) [inline features/step_definitions/application_steps.rb]: This step def is an assertion. . I would place it in assertion_steps

**Review by tochman — CHANGES_REQUESTED** [review]

**Comment by tochman** [comment]
Please avoid commit messages like `for the hound` They are not very descriptive.

File: db/migrate/20161112171216_add_status_to_membership_application.rb
Comment (diraulo) [reply db/migrate/20161112171216_add_status_to_membership_application.rb:1]: Ignore this one...

File: config/routes.rb
Comment (thesuss) [reply config/routes.rb]: Which PR/feature is that? It (the update method in controller and route) is there because I put it there in order for this feature to pass its tests, if it is incorrect the feature was not described well enough.

Comment (tochman) [reply config/routes.rb]: Well, I would say the Edit Application feature.

Comment (thesuss) [reply config/routes.rb]: So i.e. this one had to wait until that one was done. 
I thought they could be done side by side. This one is paused until that one is in develop and any merge conflicts will be resolved then.

Comment (tochman) [reply config/routes.rb]: Okay. Thx

File: features/membership_application_status.feature
Comment (tochman) [inline features/membership_application_status.feature]: I would suggest renaming this feature to `update_membership_application_status.feature`

File: features/step_definitions/assertion_steps.rb
Comment (tochman) [inline features/step_definitions/assertion_steps.rb:29]: This step mixes actions and assertions. It will return a false positive imo

File: features/step_definitions/navigation_steps.rb
Comment (tochman) [inline features/step_definitions/navigation_steps.rb]: Imo this step def should be in `membership_application_steps.rb`

File: features/update_membership_application.feature
Comment (tochman) [inline features/update_membership_application.feature:10]: The only users used in this test are 

```
 | din@mail.se            |       |
 | admin@sgf.com     | true  |
```

No need to keep `applicant_1@random.com` around

File: features/step_definitions/basic_steps.rb
Comment (tochman) [inline features/step_definitions/basic_steps.rb]: I think you can omit `page`

File: app/views/memberships/show.html.haml
Comment (tochman) [inline app/views/memberships/show.html.haml]: A bit too much logic in the view imo. I would suggest extracting this to a helper class.

File: features/step_definitions/assertion_steps.rb
Comment (thesuss) [reply features/step_definitions/assertion_steps.rb:29]: I don't think it does. It returned Pending until I got the logic on the view working (which yes, is to be moved/refactored)

File: app/views/memberships/show.html.haml
Comment (thesuss) [reply app/views/memberships/show.html.haml]: I agree, just haven't gotten to it (WIP)

Comment (tochman) [reply app/views/memberships/show.html.haml]: Cool, just giving feedback. ;-)

File: features/step_definitions/assertion_steps.rb
Comment (tochman) [inline features/step_definitions/assertion_steps.rb]: Well. if you perform an action in a step that implies in the title that it is an assertion, that will always fise a red flag with me. I would like to get a second opinion on this though.

Comment (tochman) [reply features/step_definitions/assertion_steps.rb:29]: Okay, after a second read I concur.

File: app/controllers/memberships_controller.rb
Comment (tochman) [inline app/controllers/memberships_controller.rb]: I'm leaning toward updating status in a custom method/route. Meaning we should probably use a custom form (`form_tag`) The `edit/update` paths should be reserved for editing application by whoever submits it (the applicant)

File: features/update_membership_application.feature
Comment (tochman) [inline features/update_membership_application.feature:35]: the reason I suggested `_status` at the end of this feature file is because it reflects the scope of this feature better.

File: Gemfile
Comment (tochman) [inline Gemfile]: Old syntax. After all this time? Still? 🤔

File: features/step_definitions/basic_steps.rb
Comment (tochman) [inline features/step_definitions/basic_steps.rb]: I would use a more generic variable name. Perhaps 'option'?

File: features/step_definitions/navigation_steps.rb
Comment (tochman) [inline features/step_definitions/navigation_steps.rb]: Still a step definition concerning MembershipApplication. Should be extracted to relevant file.

File: features/update_membership_application.feature
Comment (tochman) [inline features/update_membership_application.feature:17]: Repeated steps can be placed in Background to keep feature more DRY

File: features/update_membership_application.feature
Comment (tochman) [inline features/update_membership_application.feature]: Same here

File: app/controllers/memberships_controller.rb
Comment (tochman) [inline app/controllers/memberships_controller.rb]: This won't work since we need to redirect to the #show page after a status update.

Comment (thesuss) [reply app/controllers/memberships_controller.rb]: But if we take the update-status and makes it a completely different route then it should not be affected by what is in "update"?

**Comment by weedySeaDragon** [comment]
There are a few things going on here:
-  The 'show' page is only working for an admin
  -  Don't use the "show" page only for an admin user.  If you want to restrict what appears on the 'show' page based on the current Pundit policy (ex: if the user is an admin or not), then set up a policy for specific attributes based on the user.
- After a user does a successful 'edit', then the user should see the _show_ page.  That's the standard RESTful way to do things.  

IOW,  the 'show' and 'edit' pages are being misused to try to get the _admin can change the status_ functionality working.  Instead, we need to use Pundit Policies to determine who can do what, and then display the appropriate thing (partial) on the view.

It makes sense to me that a user can see the status of their application.  (A user will want to know: "Is it still pending?" ).  But a user cannot change (edit) the status of their application.  Only a SHF administrator can do that.  Thus we need a Pundit policy that says "if a user is an admin, then the user can change (edit) the MembershipApplication status.  Otherwise, the user can only view the status."  And so on the view, if the policy says it's ok to edit, we'll show a _form_.  Otherwise, we'll just show the status.

**Comment by weedySeaDragon** [comment]
This is still more complicated than it needs to be.  

Rails conventions are not being followed:  
- after a MembershipApplication has been edited (with the _edit_ action), then the user should see the _show_ view.  That is the Rails convention.
- the _show_ view is not being used as it should be
- _show_ is a standard view:  The ApplicationPolicy must have a rule for the _show view_
  - if you want to have an _additional_ ability specifically for the admin user, then _add_ that information to the MembershipApplication policy; don't clobber the standard _show_ action.  (And limit it to the MembershipApplicationPolicy.  It does not apply to the entire system (ApplicationPolicy) )

**Pull request closed without merging**
